src/Pillar-ExporterCore/PRTarget.class.st

```diff
@@ -8,6 +8,46 @@ Class {
 	#category : #'Pillar-ExporterCore-Base'
 }
 
+{ #category : #clap }
+PRTarget class >> buildClapCommand [
+	
+	| target project |
+	target := self new.
+	project := self newProject.
+ 
+	^ (ClapCommand id: self clapId)
+		description: 'Build your Pillar documents in ', self clapId ,' format';
+		add: ClapFlag forHelp;
+		add: ((ClapFlag id: #templatable)
+				description: 'Use a template');
+				meaning: [project beTemplatable];
+		add: ((ClapFlag id: #all)
+				description: 'Build all Pillar documents in repository';
+				canonicalName: 'a';
+				meaning: [ PRBuildAllStrategy new ]);
+		add: ((ClapFlag id: #mainRoot)
+				description: 'Build index.pillar in root directory';
+				canonicalName: 'm';
+				meaning: [ PRBuildRootMainStrategy new ]);
+		add: ((ClapPositional id: #requestedFiles)
+				description: 'Pillar files you want to build';
+				multiple: true;
+				meaning: [ :doc | doc word asFileReference ]);
+		meaning: [ :args |
+			args at: #helpFlag ifPresent: [ :help | help value; exitSuccess ].
+			args at: #templatable ifPresent: [(args at: #templatable) value ].
+			"here, default strategy is mainRoot and the following order give strategies a priority order"
+			(args at: #requestedFiles ) isExplicit ifTrue:[ 
+				target buildStrategy: (PRBuildListStrategy list: (args occurrencesOf: #requestedFiles collect: #value) )
+				].
+			args at: #all ifPresent: [ :strat | target buildStrategy: strat value ].
+			args at: #mainRoot ifPresent: [ :strat | target buildStrategy: strat value].
+			
```

Comment:
So the meaning of the flags does not execute automatically? :(
Why is that? Couldn't we have them executed automatically?

Like that The ClapCommands could marry well with a builder pattern.

@cdlm ?

Reply:
If flags have side-effects then their order of execution must be clear; however there is no obvious order that would be valid in all situations. Sometimes the order of declaration in the specification makes sense, sometimes it's the order of invocation on the command line, and it could also be some arbitrary order depending on the application. Some flags can also be repeated.

Clap can not make an assumption how and when the app will use its flags. So they are best seen as simple parameters that you query instead of an imperative API to a statful builder. The meaning blocks are there to lift the string passed on the command line into an object value that's useful to the app.

For commands the situation is different because each subcommand selects a narrower subset of the app's behavior. In the end you have one and only one most-specific subcommand, so that makes a convenient entry point.

Reply:
> If flags have side-effects then their order of execution must be clear

Yes, and Clap could define an order :). Even more, it could have a default one "definition order" for example. And then accept other strategies like "invocation order", and each command could define the order of flags with a strategy, no?

> I didn't want to assume how and when an app should use its flags. So they are best seen as optional parameters that you use when you need their value, instead of an imperative API to a stateful builder. The meaning blocks are there to lift the string passed on the command line into an object value that's useful to the app.

I understand. But to me, we should take a stand and favour one usage over the other :).
And maybe give hooks to the user to override those decisions.
We will never make a choice that pleases everybody, I'd take that outside the table :P.
My heuristic here is that we should favour most common usages, and leave the freedom to do less common usages.

We are now using a full-grown framework to parse command line arguments, so we should extract the juice from it, no?

I'd like to discuss alternatives, tell me what you think:

1. **flags with effect**. The command executes the flag meaning automatically. This works naturally with optional flags. If some flag is not there, the building method is not called.

```smalltalk
myBuilder := MyBuilder new.
ClapCommand new
   add: ((ClapFlag id: #x) meaning: [ myBuilder withX ]);
   add: ((ClapFlag id: #y)
      beIntegerValue; "clap does the parsing and validation for me ;)"
      meaning: [ :anInteger | myBuilder withY: anInteger ]);
   add: ((ClapFlag id: #z) meaning: [ :aString | myBuilder withZ: aString asUppercase ]);
   meaning: [ myBuilder build ].
```

2. **flags with value as explicit argument**. The command executes the flag automatically and sends the associated values as argument of the command meaning. Here the meaning of flags defines actually how to "marshall" the flag, what is their value when used from the command line (otherwise having meaning have no sense, no?...).
=> But, as a side effect, all flags become mandatory, because they are arguments in the block.

```smalltalk
ClapCommand new
   add: ((ClapFlag id: #x) meaning: [ SomeObject new ]);
   add: ((ClapFlag id: #y) meaning: [ 17 ]);
   add: ((ClapFlag id: #z) meaning: [ :value | value asInteger ]);
   meaning: [ :x :y :z | MyObject withX: x withY: y withZ: z ].
```

3. **flags with value in dictionary**. The command executes the flag automatically and just stores the associated values in a dictionary. Here the meaning of flags, again, defines how to "marshall" the flag, what is their value when used from the command line.
In this version, what we have now IIUC, it's the user responsibility to have conditional code to check for the arguments, which I don't like...

```smalltalk
ClapCommand new
   add: ((ClapFlag id: #x) meaning: [ SomeObject new ]);
   add: ((ClapFlag id: #y) meaning: [ 17 ]);
   add: ((ClapFlag id: #z) meaning: [ :value | value asInteger ]);
   meaning: [ :args | MyObject withX: (args at: #x) withY: (args at: #y) withZ: (args at: #z ifAbsent: [0]) ].
```

I prefer solution 1. I think it is easy to understand and that defining a clear order of execution of flags can simplify the task of the developer (I think definition order is the best one because the developer can foresee what will happen more easily).

Solution 2 is too limiting, and solution 3 too verbose.

But please note that solutions 1 and 3 are somewhat compatible ;). We could use solution 1 and, those people wanting fine grained control could use solution 3.

> We tried a version where the block had a side-effect, but then in the meaning block it looked like `(args at: #all) value.` with no immediate indication of what effect this has.

I did not understand this. My point is that nobody should call `(args at: #all) value.` at all :). It seems superfluous. It should be called automatically.
Otherwise, why should I do `value`? 

```smalltalk
ClapCommand new
   add: ((ClapFlag id: #x) meaning: [ SomeObject new ]);
   meaning: [ :args | (args at: #x) value ].
```

If the evaluation could be done automatically when accessing the argument:

```smalltalk
ClapCommand new
   add: ((ClapFlag id: #x) meaning: [ SomeObject new ]);
   meaning: [ :args | args at: #x ].
```

or even simpler, because people do not need to understand the semantics of `meaning`:

```smalltalk
ClapCommand new
   add: (ClapFlag id: #x);
   meaning: [ :args | args at: #x ifPresent: [ SomeObject new ] ].
```

I don't know, tell me what you think about these ideas. I am really convinced that clap could do more :)

Reply:
I'd like to move that discussion somewhere more specific to Clap itself, if you don't mind. I tried adding a wiki page but that doesn't really seem to lend itself to a discussion format. An issue could work. What do you think?

Reply:
Sure, do you want that I create an issue?

Reply:
Done, see https://github.com/cdlm/clap-st/issues/29